### PR TITLE
OpenGL: Allow multiple glsl() stages

### PR DESCRIPTION
**Pull request by shoaibkamil**
This fixes the bug reported by @estollnitz and allows multiple stages to be scheduled `compute_root().glsl()`.

This is sub-optimal in terms of performance in the following ways:
- currently we use a linked list to track the different kernels
- we use `strcmp()` to match the called kernel with the right compiled kernel

PTAL & if there's any best practices for GPU runtimes w.r.t. the above sub-optimality please advise.

### CODE REVIEW

File: src/runtime/opengl.cpp [hunk 1]
```diff
@@ -336,6 +337,17 @@ WEAK char *strndup(const char *s, size_t n) {
     return p;
 }
 
+WEAK char *strstrip(char *str, size_t n) {
```
Comment (abadams) [inline src/runtime/opengl.cpp]: What does this do? Looks like it consumes a token. If so, please add a comment.

File: src/runtime/opengl.cpp [hunk 2]
```diff
@@ -336,6 +337,17 @@ WEAK char *strndup(const char *s, size_t n) {
...
+
+  char *pos = str;
+  while (pos != str+n
+         && *pos != '\n'
+         && *pos != ' ')
```
Comment (abadams) [inline src/runtime/opengl.cpp]: Might it be good to add && *pos != '\0' to the condition? Also we use braces on ifs.

File: src/runtime/opengl.cpp [hunk 3]
```diff
@@ -464,7 +476,8 @@ WEAK KernelInfo *create_kernel(void *user_context, const char *src, int size) {
 
         const char *args;
         if ((args = match_prefix(line, kernel_marker))) {
-            // ignore
+            // set name
+            kernel->name = strstrip(strndup(args, next_line - args), next_line - args);
```
Comment (abadams) [inline src/runtime/opengl.cpp]: This string looks like it's never freed. Is there somewhere modules are cleaned up? If it's intentionally leaked, please add a comment.

File: src/runtime/opengl.cpp [hunk 4]
```diff
@@ -1383,6 +1396,24 @@ WEAK int halide_opengl_copy_to_host(void *user_context, buffer_t *buf) {
 
 using namespace Halide::Runtime::Internal::OpenGL;
 
+// Find the correct module for the called function
+// TODO: This currently takes O(# of GLSL'd stages) and can
+// be optimized
+WEAK ModuleState* find_module(const char *stage_name) {
+  ModuleState* state_ptr = state_list;
```
Comment (abadams) [inline src/runtime/opengl.cpp]: We use 4-space indentation

**Comment by dsharletg** [comment]
Is this ready to merge? It looks like the one open question has been addressed.

**Comment by abadams** [comment]
It seems like it leaks memory to me for the shader sources via strndup. I'm
not sure why this needs to allocate memory at all - can't it just inject
zero bytes in between the shaders?

**Comment by dsharletg** [comment]
I think it's getting freed now: https://github.com/halide/Halide/pull/1439/files#diff-a4078cd7d310d958ac48914d7987853cR563

But yeah, I agree, it would be better if strdup isn't necessary.

**Comment by shoaibkamil** [comment]
Good point; give me a couple hours to change it to get rid of malloc/strdup. 





On Wed, Aug 17, 2016 at 12:51 PM -0400, "Dillon Sharlet" <notifications@github.com> wrote:












I think it's getting freed now: https://github.com/halide/Halide/pull/1439/files#diff-a4078cd7d310d958ac48914d7987853cR563



But yeah, I agree, it would be better if strdup isn't necessary.

**Comment by shoaibkamil** [comment]
I've addressed the copying of the shader sources and the leak associated with that.

I've left in the use of `strdup` for copying the shader name into a string; this is used during lookup and is freed in the line @dsharletg pointed to.

File: src/runtime/opengl.cpp [hunk 5]
```diff
@@ -1883,6 +1914,27 @@ WEAK int halide_opengl_device_sync(void *user_context, struct buffer_t *) {
...
+
+  const char* sentinel = strstr(src+1, kernel_marker);
+  bool done = false;
+  char *ret = (char*)src;
+
+  if (sentinel == NULL) {
+    *len = strlen(src);
+    done = true;
+  }
+  else
```
Comment (abadams) [inline src/runtime/opengl.cpp]: Please use this style:

    if (foo) {
        asd
    } else {
        asd
    }

File: src/runtime/opengl.cpp [hunk 6]
```diff
@@ -544,6 +560,7 @@ WEAK void delete_kernel(void *user_context, KernelInfo *kernel) {
         arg = next;
     }
     free(kernel->source);
```
Comment (abadams) [inline src/runtime/opengl.cpp]: It is still correct to free the source?

File: src/runtime/opengl.cpp (hunk 5 above)
Comment (abadams) [inline src/runtime/opengl.cpp]: (and four-space indentation)

File: src/runtime/opengl.cpp [hunk 7]
```diff
@@ -1383,6 +1400,23 @@ WEAK int halide_opengl_copy_to_host(void *user_context, buffer_t *buf) {
...
 
+// Find the correct module for the called function
+// TODO: This currently takes O(# of GLSL'd stages) and can
+// be optimized
+WEAK ModuleState* find_module(const char *stage_name) {
+    ModuleState* state_ptr = state_list;
+
+    while (state_ptr != NULL) {
+        KernelInfo *kernel = state_ptr->kernel;
+    if (kernel && strcmp(stage_name, kernel->name) == 0) {
```
Comment (abadams) [inline src/runtime/opengl.cpp]: Indentation seems messed up here

File: src/runtime/opengl.cpp (hunk 6 above)
Comment (shoaibkamil) [inline src/runtime/opengl.cpp]: The original implementation copies the kernel source into the `kernel` struct; previously, my PR was adding an additional (leaking) copy.  I've now removed my extra copy, but this applies to the source copied by the original implementation.

Optionally, we could just not store the source.  Let me know if you want me to make that change.

**Comment by shoaibkamil** [comment]
Sorry for the indentation issues; that's embarrassing.  Hopefully this is now ready to merge.

**Pull request merged**